### ISSUE TITLE
TensorBoard: Refresh Backend when the frontend refreshes

### ISSUE THREAD

**Issue by s-gv** [tensorflow/tensorflow#2050]
Tensorboard seems to scan the log infrequently. Right now, I have to kill and re-start the server to make it re-scan the log. Could we have tensorboard re-scan the log when the tensorboard webpage is refreshed?

**Comment by danmane** [comment]
It scans every minute (although it may take longer to finish loading the data). In master you can configure this via [a flag on the TensorBoard process](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/tensorboard/tensorboard.py#L60) although that didn't make it into 0.8.

It would be nice to wire up the reload button to also trigger an immediate reload on the backend. But I don't have any near-term plans to implement that.

In practice, if TensorBoard seems to be scanning the log infrequently, it is more likely that this is because you are writing to multiple summary files simultaneously (ie. you have more than one SummaryWriter active at a time). TensorBoard expects that only one file is written to at a time, so if you have an early file A and a later file B, once B is ever written to, it assumes file A is finalized and will stop reading events from file A. So if most of your actvitiy is in file A it will look like TensorBoard has just stopped updating. When you kill and restart the server, it reads everything from file A before going on to file B, so the result is it loads data that was hidden to it before.

We're currently working on code to have TensorBoard detect this case and warn you in the UI, so it should become less confusing.

**Comment by s-gv** [comment]
My scenario is a bit different. I am trying to check if my model implementation is correct by over-fitting a small training set while varying the learning rate. Because the training set is small, it completes in about a second or so. But the tensorboard visualization is stale, so I end up re-starting the server. 

I'd be happy implement triggering the log reload on refresh if you're OK with that feature. Could you please point me to where I should poke around?

**Comment by danmane** [comment]
I see; that makes sense then. Your help with adding this feature would be extremely welcome :) here's an overview of how we could do it.

Overall, TensorBoard consists of a few major pieces. On the backend, there are the [`EventMultiplexer`](https://github.com/tensorflow/tensorflow/blob/0ebbb99084091bba963f7b452d9f2fa93c70f6e2/tensorflow/python/summary/event_multiplexer.py) and [`EventAccumulator`](https://github.com/tensorflow/tensorflow/blob/0ebbb99084091bba963f7b452d9f2fa93c70f6e2/tensorflow/python/summary/event_accumulator.py), a pair of classes which provide an API for accessing summary data from TensorFlow. 

The [`Server`](https://github.com/tensorflow/tensorflow/blob/0ebbb99084091bba963f7b452d9f2fa93c70f6e2/tensorflow/tensorboard/backend/server.py) instantiates these classes and creates a `MultiplexerReloadingThread` which sets a timer to automatically reload the `Multiplexer`. If the `Multiplexer` is already caught-up to historical data, it will generally load very quickly, but if you just turned TensorBoard on and it has 5 gigs of historical data to chew through, it may take a while.

[`handler.py`](https://github.com/tensorflow/tensorflow/blob/0ebbb99084091bba963f7b452d9f2fa93c70f6e2/tensorflow/tensorboard/backend/handler.py#L460) contains all the handlers for the server's routes. The API is documented in the file [http_api.md](https://github.com/tensorflow/tensorflow/blob/0ebbb99084091bba963f7b452d9f2fa93c70f6e2/tensorflow/tensorboard/http_api.md). Please take a moment to read that (the next paragraph will assume some familiarity).

The way reloading works is as follows: the user clicks the reload button, which calls the `reload` method that is implemented by `TF.Backend.Behavior`. The behavior first reloads the run-tag mapping, and then triggers every existing data display element (e.g. a chart or image loader) to reload.

Thus, if we want to have the reload trigger a backend reload, we should have the first method that is called on the backend (the `runs` route) delay while the backend reloads. I think a good way to implement this would be to have an optional query parameter passed to the route (?reload), and if the query is present, then rather than respond immediately, the server reloads the backend and then responds.

However, as I mentioned above, the reload may take a long time if the server just turned on and there is a lot of historical data to explore. So we'll need to have a timeout, say 5s - if the backend reload takes more than 5 seconds, then we resolve early even though we aren't done loading.

It would be really nice to further return some info on whether the backend is done loading or not so we can put a little "data still loading!" type indicator, but this change is complicated enough without doing that :)

Does that seem reasonable?

**Comment by girving** [comment]
@danmane: Any chance this is obviated by improvements since then?

**Comment by danmane** [comment]
This is still a valid request. I changed the issue name to better reflect where the conversation went, though.

**Comment by gcr** [comment]
Wait, we're not supposed to write to multiple SummaryWriters in a single session? Why does the official TensorBoard "getting started" tutorial showcase this behavior if it's not supported?

I'm trying to (unsuccessfully) base my example off of https://www.tensorflow.org/versions/r0.9/how_tos/summaries_and_tensorboard/index.html , which creates a `train_writer` and a `test_writer`. Aren't they both necessary?

**Comment by danmane** [comment]
Sorry, I misspoke. You can't have multiple summary writers writing to the
same run, i.e. In the exact same subdirectories. Having multiple summary
writers pointing at different directories, as is the case in the tutorial,
is fine.

**Comment by gcr** [comment]
Ah, I see! Thanks for the clarification.

**Comment by waleedka** [comment]
Is this in the master branch now? I see a refresh icon in the top right corner of the page next to the settings gear, so I was wondering if that was the implementation that @danmane  is referring to. However, clicking that icon doesn't seem to do anything for me. 

During development I edit my models often and run quick training sessions that take a minute or less to see the effects of my edits. I delete all the events data from the model directory, but TensorBoard continues to show the old events. I have to repeatedly stop and restart the tensorboard process to see the new data. I was happy when I noticed the refresh icon, but it doesn't seem to be working for me.

**Comment by lethienhoa** [comment]
Dear all,
I run a program on the server (which is much faster because it has GPU supported). I download the summaries down to my local machine and try to visualize the result with Tensorboard. However, I encounter these 2 problems:
1. It shows training's performance but not dev's performance (though I downloaded dev's summaries in the same sub-directory as well)
2. The Tensorboard just show result of 500 first iteration steps while my experiment has >22k steps. I had to press the button "Toggle all run" many times to refresh the new result. Are there any better ways ?
I appreciate very much for any possible help.

**Comment by dandelionmane** [comment]
@lethienhoa 
1. Try putting dev's summaries in a different sub-directory.
2. Try launching TensorBoard with --reload_interval=5 or some similar smaller value. Then pushing the refresh button (in top right corner) should get you new data more regularly.

**Comment by lethienhoa** [comment]
Thank you very much for your answer. It worked for the first point ! However, on the second point, though I've tried --reload_interval=1 (minimum int value for this var) but it doesn't speed up too much. I still have to push the refresh button thousands of time...
Do you have any other better ways ?

**Comment by MInner** [comment]
Try putting `setInterval(function() {document.getElementById('reload-button').click()}, 3000);` into chrome console :)

**Comment by liufuyang** [comment]
Another question, can the tensorboard auto refresh the page when new logs are generated? (Perhaps not according to the comments above...) 
And is there a plan to implement something like this?

**Comment by lethienhoa** [comment]
Hi,
I've figured that we don't need to do any refresh page by hand at all. When we open the terminal by "tensorboard --logdir yoursummarydirectory", open up your browser and wait (~5 or 10 min depending on the size of your model) and it will automatically load all.
I don't know why there is a delay like that but you definitely don't need to do boring stuffs anymore, just wait !

**Comment by liufuyang** [comment]
@lethienhoa I think you issues is more like "loading slow". But for me I hope to monitoring the summary metrics on the fly while the calculation is running generating logs. Currently the page won't refresh itself but MInner's solution above seems very handy at the moment.

**Comment by jart** [comment]
This issue has been migrated to https://github.com/tensorflow/tensorboard/issues/80.

**Comment by Eugenio-Bruno** [comment]
> @lethienhoa , try putting `setInterval(function() {document.getElementById('reload-button').click()}, 7000);` into chrome console `Ctrl+Shift+J` :)

If you are getting here from google, and the above gives you an error, it seems that you might have to change that code into this (I don't understand javascript but I did a bit of googling):

```
setInterval(
  function() {
    document.getElementsByTagName("tf-tensorboard")[0].shadowRoot.getElementById("reload-button").click()
  }, 7000
);
```

**Comment by medinanicolas** [comment]
> > @lethienhoa , try putting `setInterval(function() {document.getElementById('reload-button').click()}, 7000);` into chrome console `Ctrl+Shift+J` :)
> 
> If you are getting here from google, and the above gives you an error, it seems that you might have to change that code into this (I don't understand javascript but I did a bit of googling):
> 
> ```
> setInterval(
>   function() {
>     document.getElementsByTagName("tf-tensorboard")[0].shadowRoot.getElementById("reload-button").click()
>   }, 7000
> );
> ```

Or simple this:

`setInterval(function() {document.getElementsByClassName('reload-button')[0].click()}, 7000);
`
